feat: ROOT-11: Support reading JSONL from source cloud storages

matt-bernstein opened the pull request:
[x] Factor out JSON parsing and validation logic from import storages into a function that can be hot-swapped in LSO/LSE
[x] add it to settings as an app-specific variable
[x] replace stdlib parsing with pyarrow parsing to handle JSONL as well as JSON
[x] test coverage for JSONL
[x] feature flag
[x] unskip multitask import tests in LSO, to prepare for testing different behavior in LSE and LSO

Implementation notes:
- had to move computation of row_index and row_group into get_data so that load_tasks_json, which will be swapped in LSO/LSE, will have the correct type in all cases - makes for a pretty large diff
- simplified error messages significantly since the relevant logic is shared/deduped
- had to stick with stdlib json parsing instead of removing it entirely:
  - in the case of single dict, pyarrow can't distinguish between this and a one-line JSONL file, so need stdlib json to set row_index=None
  - in the case of top-level list, pyarrow can't parse this at all since it assumes each line is an entry in a table. There is an [open issue](https://github.com/apache/arrow/issues/46486) for this.
  - consider peeking at file extensions to avoid double-parsing (for clarity, not speed) - see TODO comment

<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

matt-bernstein committed:
feat: ROOT-1: Add offset to storage link model to support storage link to task within file of tasks on cloud storage

matt-bernstein committed:
review comments

matt-bernstein committed:
remove whitespace

Co-authored-by: Jo Booth <jo.m.booth@gmail.com>

matt-bernstein committed:
update comment

Co-authored-by: Jo Booth <jo.m.booth@gmail.com>

matt-bernstein committed:
Merge branch 'develop' into 'fb-ROOT-1'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15027678115

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15027678115

matt-bernstein committed:
feat: ROOT-9: Allow reading multiple tasks from a JSON file in source cloud storage

matt-bernstein committed:
update comment

matt-bernstein committed:
update storagelink creation

matt-bernstein committed:
fix DM action

matt-bernstein committed:
blue

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-9

matt-bernstein committed:
udpate test

matt-bernstein committed:
futureproof create method

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15030860401

matt-bernstein committed:
Merge branch 'develop' into 'fb-ROOT-9'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15045532193

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15045586529

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-9

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15052428789

matt-bernstein committed:
make row_index=none default

matt-bernstein committed:
get_data return type

matt-bernstein committed:
blue

matt-bernstein committed:
test new model fields

matt-bernstein committed:
feat: ROOT-11: Support reading JSONL from source cloud storages

matt-bernstein committed:
split out repeated json parsing logic from import storages

robot-ci-heartex committed:
Apply pre-commit linters

matt-bernstein committed:
add to settings

matt-bernstein committed:
comment

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-11

matt-bernstein committed:
add pyarrow lib

matt-bernstein committed:
cleanup GCS utils

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15074933586)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15074933586

matt-bernstein committed:
Merge branch 'develop' into 'fb-ROOT-11'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15074933586

matt-bernstein committed:
fix gcs

matt-bernstein committed:
organize import

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-11

matt-bernstein committed:
handle localfiles

matt-bernstein committed:
remove unused TaskValidationError

matt-bernstein committed:
add jsonl; compute row_idx and row_group in get_data instead of afterwards

matt-bernstein committed:
don't pass storage class name

matt-bernstein committed:
more permissive parsing

matt-bernstein committed:
fix recursion

jombooth commented on label_studio/io_storages/azure_blob/models.py:
```diff
@@ -209,28 +209,16 @@ def iterkeys(self):
             yield file.name
 
-    def get_data(self, key) -> Union[dict, list[dict]]:
+    def get_data(self, key) -> tuple[list[dict], list[int | None], list[int | None]]:
```
Once we're at a 3-tuple return type, I think it'd be better to create a simple object with a constructor and return that, or use a solution like NamedTuple or TypedDict. Python errors from unpacking tuples into too few variables in assignment statements are quite common

jombooth replied:
also, for the `list[int | None]` types, I think it'd be better to say `list[int] | Literal[[None]]`, since there is no other list containing `None`s that we could want returned afaik.

matt-bernstein replied:
Not a bad idea. I'm thinking `(task_data: list[dict] (or pa.Table), storagelink_params: list[StorageLinkParam])`

matt-bernstein replied:
though tbf, this function will only ever have one caller

jombooth replied:
I also think that perhaps we should find an alternative to returning lists containing a single `[None]` - I get why we have them now, but it seems to me we could find a better way of representing this concept in Python

matt-bernstein replied:
done

jombooth commented on label_studio/io_storages/utils.py:
```diff
@@ -109,3 +116,61 @@ def parse_range(range_header):
+        raise ValueError(
+            (
+                f"Can't import JSON-formatted tasks from {key}. If you're trying to import binary objects, "
+                f'perhaps you\'ve forgot to enable "Treat every bucket object as a source file" option?'
```
```suggestion
                f'perhaps you forgot to enable "Treat every bucket object as a source file" option?'
```

jombooth commented on label_studio/io_storages/utils.py:
```diff
@@ -109,3 +116,61 @@ def parse_range(range_header):
+    def _error_wrapper(exc: Optional[Exception] = None):
+        raise ValueError(
+            (
+                f"Can't import JSON-formatted tasks from {key}. If you're trying to import binary objects, "
```
```suggestion
                f"Can’t import JSON-formatted tasks from {key}. If you’re trying to import binary objects, "
```

jombooth commented on label_studio/io_storages/utils.py:
```diff
@@ -109,3 +116,61 @@ def parse_range(range_header):
+            )
+        ) from exc
+
+    # TODO: rely on file extensions here instead of pure exception-chaining? Would be more readable, potentially less reliable, would work for all storage types except redis.
```
I think exception chaining is actually a better answer than file extensions, because it's not uncommon to use .txt or similar to represent JSON. Would favor removing this comment

jombooth commented on label_studio/io_storages/gcs/models.py:
```diff
@@ -180,28 +180,17 @@ def iterkeys(self):
             bucket_name=self.bucket,
             key=key,
-            convert_to=GCS.ConvertBlobTo.JSON,
+            convert_to=GCS.ConvertBlobTo.NOTHING,
```
do we have to pass `convert_to` at all in this case?

matt-bernstein replied:
Nope! Just tried to be very explicit. But it's not needed, will remove it

jombooth commented on label_studio/io_storages/utils.py:
```diff
@@ -109,3 +116,61 @@ def parse_range(range_header):
+
+    # TODO: rely on file extensions here instead of pure exception-chaining? Would be more readable, potentially less reliable, would work for all storage types except redis.
+    try:
+        value = json.loads(blob_str)
+    except json.decoder.JSONDecodeError as e:
+        if flag_set('fflag_feat_root_11_support_jsonl_cloud_storage'):
+            try:
+                table = pyarrow.json.read_json(
+                    pa.py_buffer(blob_str), parse_options=pa.json.ParseOptions(newlines_in_values=True)
+                )
```
hmm, seems to me we may want to handle this differently - as written, with the FF enabled we have to fail to parse JSON to even attempt parsing JSONL. Could we condition whether we try to `json.loads()` vs `pyarrow.json.read_json()` on whether the FF is on or not? It might be okay as is, if we're confident that `json.loads()` will fail quickly on a large JSONL file.

matt-bernstein replied:
`json.loads()` will fail on the first character of the second line of a JSONL file:

```
JSONDecodeError                           Traceback (most recent call last)
Cell In[1], line 1
----> 1 json.loads(b'{"a":1}\n{"a":2}')

File ~/anaconda3/lib/python3.12/json/__init__.py:346, in loads(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    341     s = s.decode(detect_encoding(s), 'surrogatepass')
    343 if (cls is None and object_hook is None and
    344         parse_int is None and parse_float is None and
    345         parse_constant is None and object_pairs_hook is None and not kw):
--> 346     return _default_decoder.decode(s)
    347 if cls is None:
    348     cls = JSONDecoder

File ~/anaconda3/lib/python3.12/json/decoder.py:341, in JSONDecoder.decode(self, s, _w)
    339 end = _w(s, end).end()
    340 if end != len(s):
--> 341     raise JSONDecodeError("Extra data", s, end)
    342 return obj

JSONDecodeError: Extra data: line 2 column 1 (char 8)
```

so it's linear in the size of the first task, not the number of tasks. Will verify this, though, I see the concern

matt-bernstein replied:
reasonably confident in ^ after reading through stdlib json source

matt-bernstein committed:
structured return type for get_data

matt-bernstein committed:
ruff

matt-bernstein committed:
string fmt

Co-authored-by: Jo Booth <jo.m.booth@gmail.com>

matt-bernstein committed:
string fmt

Co-authored-by: Jo Booth <jo.m.booth@gmail.com>

matt-bernstein committed:
remove comment

matt-bernstein committed:
remove option

matt-bernstein committed:
bugfix

matt-bernstein committed:
change structured type, add tests

matt-bernstein committed:
wip fix ff mocking

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-11

matt-bernstein committed:
wip convert to pytest

jombooth commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15147464154)

jombooth committed:
regen lockfile

matt-bernstein committed:
fix tests

jombooth committed:
place django_db mark at file level

matt-bernstein committed:
handle None

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15148947939)

matt-bernstein committed:
wip mock ffs

matt-bernstein committed:
ensure blob_str enters as bytes

matt-bernstein committed:
add debug log, fix azure test

matt-bernstein committed:
fix imputed field in task from other tasks in file

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-11

matt-bernstein committed:
fix redis

matt-bernstein committed:
fix mixed task formats and tests

jombooth commented on label_studio/core/settings/base.py:
```diff
@@ -598,6 +598,7 @@
 MEMBER_PERM = 'core.api_permissions.MemberHasOwnerPermission'
 RECALCULATE_ALL_STATS = None
 GET_STORAGE_LIST = 'io_storages.functions.get_storage_list'
+STORAGE_LOAD_TASKS_JSON = 'io_storages.utils._load_tasks_json'
```
Usually the meaning of a function or method that starts with a `_` is "this should not be used outside of this class/module/whatever" - seems to me that we are treating this function as more of a public one, since it's being referred to in this other file. Would it be a pain to rename?

jombooth commented on label_studio/io_storages/utils.py:
```diff
@@ -109,3 +116,73 @@ def parse_range(range_header):
+
+
+@dataclass
+class StorageObjectParams:
```
I think `StorageObject` would be a better name for this, since it actually contains the task now

jombooth commented on label_studio/io_storages/utils.py:
```diff
@@ -109,3 +116,73 @@ def parse_range(range_header):
+
+    @classmethod
+    def bulk_create(
+        cls, task_datas: list[dict], key, row_idxs: list[int] | None = None, row_groups: list[int] | None = None
```
I would prefer `row_indexes` - generally in favor of avoiding abbreviations in variable names unless they're truly ubiquitous.

matt-bernstein commented:
Note for posterity: `pa.Table.from_pylist` has a footgun in that it only constructs the schema from the first row; use `from_json` with BytesIO next time since it constructs the schema from all rows

robot-ci-heartex committed:
Apply pre-commit linters

jombooth commented on label_studio/io_storages/utils.py:
```diff
@@ -109,3 +116,73 @@ def parse_range(range_header):
+        key (str): The key of the blob. Used for error messages.
+
+    Returns:
+        list[dict]: parsed tasks.
```
```suggestion
```

matt-bernstein committed:
fix docstring

Co-authored-by: Jo Booth <jo.m.booth@gmail.com>

jombooth commented on pyproject.toml:
```diff
@@ -49,6 +49,7 @@ dependencies = [
     "ordered-set (==4.0.2)",
     "pandas (>=2.2.3)",
     "psycopg2-binary (==2.9.10)",
+    "pyarrow (>=18.0.0,<19.0.0)",
```
```suggestion
    "pyarrow (>=18.1.0,<19.0.0)",
```

Would recommend we don't allow people to go older than the version in poetry.lock, note this will require a relock

jombooth commented on label_studio/io_storages/utils.py:
```diff
@@ -109,3 +116,73 @@ def parse_range(range_header):
+    _error_wrapper()
+
+
+def load_tasks_json(blob_str: str, key: str) -> tuple[list[dict], list[StorageObjectParams]]:
```
```suggestion
def load_tasks_json(blob_str: str, key: str) -> list[StorageObjectParams]:
```

jombooth commented on label_studio/io_storages/base_models.py:
```diff
@@ -515,6 +523,7 @@ def sync(self):
                 self.info_set_queued()
                 import_sync_background(self.__class__, self.id)
             except Exception:
+                logger.debug(f'Storage {self} failed', exc_info=True)
```
```suggestion
                # needed to facilitate debugging storage-related testcases, since otherwise no exception is logged
                logger.debug(f'Storage {self} failed', exc_info=True)
```

jombooth replied:
This is a rare case where I'm in favor of a comment - the reason it's here is fairly subtle and not at all obvious from the context

matt-bernstein committed:
rename symbols

jombooth commented on label_studio/io_storages/base_models.py:
```diff
@@ -431,7 +438,8 @@ def _scan_and_create_links(self, link_class):
             try:
-                tasks_data = self.get_data(key)
+                # list of (task data + ImportStorageLink details)
+                links_params = self.get_data(key)
```
with the change to `StorageObject`, this can be `storage_objects`, which I think will be a lot more readable than `links_params` vs `link_params` (although, valiant effort at making the plurals situation readable enough!)

jombooth commented on label_studio/io_storages/base_models.py:
```diff
@@ -431,7 +438,8 @@ def _scan_and_create_links(self, link_class):
             logger.debug(f'{self}: found new key {key}')
             try:
-                tasks_data = self.get_data(key)
+                # list of (task data + ImportStorageLink details)
```
```suggestion
```

matt-bernstein committed:
add comment

Co-authored-by: Jo Booth <jo.m.booth@gmail.com>

matt-bernstein committed:
remove comment

Co-authored-by: Jo Booth <jo.m.booth@gmail.com>

jombooth commented on label_studio/io_storages/base_models.py:
```diff
@@ -341,17 +342,20 @@ def _scan_and_create_links_v2(self):
-    def add_task(cls, data, project, maximum_annotations, max_inner_id, storage, key, row_index, link_class):
+    def add_task(cls, project, maximum_annotations, max_inner_id, storage, link_params, link_class):
+        link_kwargs = asdict(link_params)
+        data = link_kwargs.pop('task_data')
```
is there a risk of a crash if the key is missing from the dict? genuinely asking, not extremely familiar with dataclasses

matt-bernstein replied:
nope, at this point it's a regular dict and `.pop` has an implicit default `None`

jombooth replied:
It's counterintuitive next to `dict#get`, but:

```
>>> {}.pop('foo')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'foo'
```

you have to pass `pop()` the `None` default if that's the expected behavior.

matt-bernstein replied:
oops, you're right!

matt-bernstein replied:
I don't think the dataclass could have had it missing to begin with, but good if someone accidentally passes another kind of object in there

matt-bernstein committed:
update min version

matt-bernstein committed:
rename

matt-bernstein committed:
defensive fix

robot-ci-heartex committed:
Apply pre-commit linters

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-11

matt-bernstein committed:
lockfile

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15188365408)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15188365408

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15192083566)
> Successfully merged:  1 file changed, 47 insertions(+), 2 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-ROOT-11'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15192083566

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15192136431)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15192136431

matt-bernstein commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15193781619)
> Successfully merged:  5 files changed, 44 insertions(+), 38 deletions(-)

matt-bernstein committed:
Merge branch 'develop' into 'fb-ROOT-11'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15193781619

jombooth committed:
add some debugging prints

jombooth committed:
more logging, add close call

jombooth committed:
more logging

jombooth committed:
Revert "more logging"

This reverts commit b2484af1318a3de2fa479c4966befcf954aeace6.

jombooth committed:
Revert "more logging, add close call"

This reverts commit 8f9dee8712126b12bd0c8de4cf56dfeaadca8ac7.

jombooth committed:
Revert "add some debugging prints"

This reverts commit 6320d8a6127bc506732d5feee50760ec83b47014.

matt-bernstein committed:
manually toggle logging FF

matt-bernstein committed:
conditionally skip test

matt-bernstein committed:
Revert "manually toggle logging FF"

This reverts commit 3d3c8e454a8ceefc24178832fbeba051c812fa29.

matt-bernstein committed:
Merge remote-tracking branch 'origin/develop' into fb-ROOT-11

matt-bernstein commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15210874380)

matt-bernstein committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15210874380
